### ISSUE TITLE
Warn against dead links within the document

### ISSUE THREAD

**Issue by ben-clayton** [DavidAnson/markdownlint#253]
It would be fantastic if bad anchor-style links were detected and warned.
I wouldn't expect bad external URL links to be caught.

**Comment by koppor** [comment]
Dead external links can be discovered with [markdown-link-check](https://github.com/tcort/markdown-link-check):

One file: `npx markdown-link-check README.md`.

Multiple files outputting the result into a text file:

```bash
find . -name \*.md -exec npx markdown-link-check -qq {} \; > bad-links.txt
```

Note that I used [npx](https://www.npmjs.com/package/npx) to avoid explicit global installation of the markdown-link-check package.

**Comment by nichtich** [comment]
Thanks for the pointer to markdown-link-check, but I'd prefer no have basic support as part of markdownlint instead of having to use yet another tool with different usage and a lot of functionality I don't need anyway. The dead-links use case in this issue is anchor-style links, so:

* extract all heading titles and [create anchor ids](https://github.com/markedjs/marked/blob/743ec55fe16e0b829f68db0b6b3cde9c0b23e78a/src/Slugger.js#L12)
* extract all link targets starting with `#` and check whether a matching id exists.

Note this would not catch link targets within inline HTML but this is [discouraged by MD033](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md033) anyway.

**Comment by DavidAnson** [comment]
@nichtich Is the anchor ID algorithm part of a specification anywhere, or are you suggesting making this specific to GitHub?

**Comment by nichtich** [comment]
Anchor links from header titles [have been discussed in CommonMark](https://talk.commonmark.org/t/feature-request-automatically-generated-ids-for-headers/) but it's not part of CommonMark. Popular implementations that create anchor ids from header titles [include Pandoc](https://pandoc.org/MANUAL.html#headings-and-sections) and GitHub. Pandoc also supports GitHub algorithm (see feature `gfm_auto_identifiers`), so GitHub algorithm seems the best choice:

* spaces are converted to dashes (`-`)
* uppercase characters are converted to lowercase characters
* punctuation characters other than `-` and `_` are removed
* if nothing is left, use the identifier `section`

For a more extensive comparison of algorithms see <https://babelmark.github.io/>

**Comment by DavidAnson** [comment]
Great info, thank you!
